src/libaktualizr/primary/aktualizr_fullostree_test.cc

```diff
@@ -0,0 +1,167 @@
+  conf.provision.provision_path = "tests/test_data/cred.zip";
+  conf.tls.server = server;
+
+  LOG_INFO << "conf: " << conf;
+
+  {
+    Aktualizr aktualizr(conf);
+
+    aktualizr.Initialize();
+
+    result::UpdateCheck update_result = aktualizr.CheckUpdates().get();
+    ASSERT_EQ(update_result.status, result::UpdateStatus::kUpdatesAvailable);
+
+    result::Download download_result = aktualizr.Download(update_result.updates).get();
+    EXPECT_EQ(download_result.status, result::DownloadStatus::kSuccess);
+
+    result::Install install_result = aktualizr.Install(update_result.updates).get();
+    EXPECT_EQ(install_result.ecu_reports.size(), 1);
+    EXPECT_EQ(install_result.ecu_reports[0].install_res.result_code.num_code,
+              data::ResultCode::Numeric::kNeedCompletion);
+  }
+
+  // do "reboot" and finalize
+  ostree_deployment.serial = 1;
+  ostree_deployment.rev = new_rev;
+  boost::filesystem::remove(conf.bootloader.reboot_sentinel_dir / conf.bootloader.reboot_sentinel_name);
+
+  {
+    Aktualizr aktualizr(conf);
+
+    aktualizr.Initialize();
+
+    result::UpdateCheck update_result = aktualizr.CheckUpdates().get();
+    ASSERT_EQ(update_result.status, result::UpdateStatus::kNoUpdatesAvailable);
+  }
+}
+
+#ifndef __NO_MAIN__
+int main(int argc, char **argv) {
+  ::testing::InitGoogleTest(&argc, argv);
+
+  logger_init();
+
+  if (argc != 3) {
+    std::cerr << "Error: " << argv[0] << " requires the path to the aktualizr-repo utility "
+              << "and an OStree sysroot\n";
+    return EXIT_FAILURE;
+  }
+  aktualizr_repo_path = argv[1];
+
+  Process ostree("ostree");
+
+  TemporaryDirectory meta_dir;
+  TemporaryDirectory temp_sysroot;
+  sysroot = temp_sysroot / "sysroot";
+  // uses cp, as boost doesn't like to copy bad symlinks
+  int res = system((std::string("cp -r ") + argv[2] + std::string(" ") + sysroot.string()).c_str());
+  if (res != 0) {
+    return -1;
+  }
+  auto r = ostree.run(
+      {"rev-parse", std::string("--repo"), (sysroot / "/ostree/repo").string(), "generate-remote/generated"});
+  if (std::get<0>(r) != 0) {
+    return -1;
+  }
+  ostree_deployment.serial = 0;
+  ostree_deployment.rev = ostree.lastStdOut();
+  boost::trim_if(ostree_deployment.rev, boost::is_any_of(" \t\r\n"));
+  LOG_INFO << "ORIG: " << ostree_deployment.rev;
+
+  std::string port = TestUtils::getFreePort();
+  server += port;
+  boost::process::child http_server_process("tests/fake_http_server/fake_test_server.py", port, "-m", meta_dir.Path());
+  TestUtils::waitForServer(server + "/");
+
+  std::string treehub_port = TestUtils::getFreePort();
+  treehub_server += treehub_port;
+  TemporaryDirectory treehub_dir;
+  boost::process::child ostree_server_process("tests/sota_tools/treehub_server.py", std::string("-p"), treehub_port,
+                                              std::string("-d"), treehub_dir.PathString(), std::string("-s0.5"),
+                                              std::string("--create"));
```

Comment:
Do you still need the `treehub_server` if you are using the `fake_test_server`? I'd hoped when I wrote that for it to be able to handle all server tasks in one script.

Reply:
I haven't tried, to be honest. This test is mostly a mix-and-match of other tests that we have.

However:
- treehub_server also generates a repository at startup which the test uses
- there is the cgi / multipart business that I don't quite understand.

Reply:
No problem. It'd be nice to clean up our various test servers, but it doesn't need to be done now. We can come to back to that later if you'd like.